### ISSUE TITLE
Feature: Import / Export for backup purposes

### ISSUE THREAD

**Issue by georgzoeller** [MohamedBassem/hoarder-app#75]
It would be nice to be able to backup export the saved content and restore it (potentially on another machine).

**Comment by MohamedBassem** [comment]
Do you want to do this per user or as a system backup? If it's a system backup, just archiving the 'DATA_DIR' is good enough. Redis data is ephemeral, and the search index can be reconstructed from the admin panel.

**Comment by streusselhirni** [comment]
For me, a system-wide backup would be enough. I think for less technical users, it would be nice to have some documentation about how to backup the `DATA_DIR`.

**Comment by talentedmrripley** [comment]
> For me, a system-wide backup would be enough. I think for less technical users, it would be nice to have some documentation about how to backup the `DATA_DIR`.

This feels like it will be specific to each users setup and their preferred backup methods and locations.  Or are you just thinking of some mention in documentation that backing up the data is enough for a restore or transfer?

**Comment by hongruilin** [comment]
> Do you want to do this per user or as a system backup? If it's a system backup, just archiving the 'DATA_DIR' is good enough. Redis data is ephemeral, and the search index can be reconstructed from the admin panel.

Hello, do you mean DATA_DIR is hoarder_data /var/lib/docker/volumes/hoarder_data/_data?

**Comment by lnicola** [comment]
It would be nice to have a better solution. I didn't check the contents of that directory, but copying data files on the fly isn't necessarily a great idea, see e.g. https://theorangeone.net/posts/database-backups/.

**Comment by kamtschatka** [comment]
You simply have to turn off your docker containers, do a backup and then restart them after the backup. I am not sure how many users your hoarder instance has, but I am guessing it should be easy enough to find a timeframe, where people are sleeping. And in the worst case, you have a 5 minute downtime.

**Comment by MRDGH2821** [comment]
Per user backup would also be nice. This can be like sqlite db file containing query results of current user :thinking:

**Comment by raviwarrier** [comment]
> Do you want to do this per user or as a system backup? If it's a system backup, just archiving the 'DATA_DIR' is good enough. Redis data is ephemeral, and the search index can be reconstructed from the admin panel.

This may seem simple for people who can and are comfortable working on the backend. But for those who are not engineers or coders - like people who use my instance of hoarder to take their backups. It would definitely improve the UX for non-technical users.

**Comment by alexkutsan** [comment]
Agree, backup and restore would be a good feature. At least as HTML file (without cached content and screenshots).

**Comment by MohamedBassem** [comment]
I have added support for a per user export of links and notes which is going to be available in the next release (https://github.com/hoarder-app/hoarder/commit/c8a3c1ee02e917b2e553d403b7bf94cbc736f51d). It's not a backup though, it's for the sake of migrating to another platform if you want to.

**Comment by raviwarrier** [comment]
That's good. Even export is fine. It just means indexing and inferencing might need to be done again, but as long as one can get a copy of all the links, it's good. :)

Keep up the good work buddy. 👍🏽

**Comment by mbuotidem** [comment]
Another use-case. I'm planning to run this on Google Cloud Run or Amazon Lightsail (container version, not the vps version) but I wouldn't feel safe without a backup mechanism in the event of container crashes. So I'd like to be able to periodically save the DATA_DIR (or just the export data) to the respective cloud's object storage. Is this something you'd be open to accepting a PR for?

**Comment by Godferdom** [comment]
It would be nice to propose an automatic (daily/monthly) backup instead of an export.

**Comment by thiswillbeyourgithub** [comment]
FYI my [karakeep-python-api](https://github.com/thiswillbeyourgithub/karakeep_python_api/) can be used to download all lists, bookmarks and highlights, etc to a text file via its API, as well as recreate it. Only per user though (the admin can't query from the other users afaik). If there's enough interest (and if @MohamedBassem does not have the time to spare to write the import/export code) I could write a script to makes at least the backup part easily. It would pair nicely with cron/systemd too. If interested create a feature request on the tool's repo: https://github.com/thiswillbeyourgithub/karakeep_python_api/

Also I think #1403 would be needed before someone else than @MohamedBassem can write a proper backup/restore script.

**Comment by MohamedBassem** [comment]
FYI, I've just released a new version of the CLI that allows you to migrate all your data between servers, and to also take a full dump of your account into an archive (including assets).

You can find the docs for the server migration here: https://docs.karakeep.app/next/guides/server-migration

This is very experimental, so use with care.

<img width="1772" height="1248" alt="Image" src="https://github.com/user-attachments/assets/a0ad9fe4-945f-489e-96bd-792b641338bb" />

**Comment by MohamedBassem** [comment]
And I'm planning to implement automatic backups as well.

**Comment by Depechie** [comment]
The CLI option works like a charm!

**Comment by juanmanavarro** [comment]
Hi.

Migration gives me this error

```
Migrating user settings …
Error: Failed migrating user settings. Reason: No "query"-procedure on path "users.settings"
✗
Error: Migration failed. Reason: No "query"-procedure on path "users.settings"
```
Any idea? 
Thanks

**Comment by CarcajadaArtificial** [comment]
> You can find the docs for the server migration here: https://docs.karakeep.app/next/guides/server-migration

This is now a broken link, the correct one is here: https://docs.karakeep.app/administration/server-migration/

**Comment by mnbf9rca** [comment]
> And I'm planning to implement automatic backups as well.

Any news on automated backups?
